snowfall/training/compute_expected_times.py

```diff
@@ -0,0 +1,144 @@
+    #
+    # TODO(fangjun): `num_repeats` is currently not used
+    phone_seqs, num_repeats = k2.ragged.unique_sequences(phone_seqs, True)
+
+    # Remove the 1st axis from `phone_seqs` (that corresponds to `seq`) and
+    # keep it for later, we'll be processing paths separately.
+    seq_to_path_shape = k2.ragged.get_layer(phone_seqs.shape(), 0)
+    path_to_seq_map = seq_to_path_shape.row_ids(1)
+
+    phone_seqs = k2.ragged.remove_axis(phone_seqs, 0)
+
+    # now compile decoding graphs corresponding to `phone_seqs` by constructing
+    # fsas from them (remember we already have the final -1's!) and composing
+    # with ctc_topo.
+    phone_fsas = _create_phone_fsas(phone_seqs)
+    phone_fsas = k2.add_epsilon_self_loops(phone_fsas)
+
+    # Set an attribute called pathphone_idx, which corresponds to the arc-index
+    # in `phone_fsas` with self-loops.
+    # Each phone has an index but there are blanks between them and at the start
+    # and end.
+    phone_fsas.pathphone_idx = torch.arange(phone_fsas.arcs.num_elements(),
+                                            dtype=torch.int32)
+
+    # Now extract the sets of paths from the lattices corresponding to each of
+    # those n-best phone sequences; these will effectively be lattices with one
+    # path but alternative alignments.
+    path_decoding_graphs = k2.compose(ctc_topo,
+                                      phone_fsas,
+                                      treat_epsilons_specially=False)
+
+    paths_lats = k2.intersect_dense(path_decoding_graphs,
+                                    dense_fsa_vec,
+                                    output_beam=10.0,
+                                    a_to_b_map=path_to_seq_map,
+                                    seqframe_idx_name='seqframe_idx')
+
+    # by seq we mean the original sequence indexes, by path we mean the indexes
+    # of the n-best paths; path_to_seq_map maps from path-index to seq-index.
+    seqs_shape = dense_fsa_vec.dense_fsa_vec.shape()
+
+    # paths_shape will also be a k2.RaggedInt with 2 axes
+    paths_shape, _ = k2.ragged.index(seqs_shape,
+                                     path_to_seq_map,
+                                     need_value_indexes=False)
+
+    seq_starts = seqs_shape.row_splits(1)[:-1]
+    path_starts = paths_shape.row_splits(1)[:-1]
+
+    # We can map from seqframe_idx  for paths, to seqframe_idx for seqs,
+    # by adding path_offsets.  path_offsets is indexed by path-index.
+    path_offsets = path_starts - k2.index(seq_starts, path_to_seq_map)
+
+    # assign new attribute 'pathframe_idx' that combines path and frame.
+    paths_lats_arc2path = k2.index(paths_lats.arcs.shape().row_ids(1),
+                                   paths_lats.arcs.shape().row_ids(2))
+
+    paths_lats.pathframe_idx = paths_lats.seqframe_idx + k2.index(
+        path_offsets, paths_lats_arc2path)
+
+    pathframe_to_pathphone = k2.create_sparse(rows=paths_lats.pathframe_idx,
```

Comment:
I am not sure whether this is correct.

Reply:
Can you perhaps print out paths_pats.pathphone_idx for the two decoding graphs and see if there is an obvious difference, e.g. perhaps one has a lot more zeros or -1's in it?

Reply:
Provided in https://github.com/k2-fsa/snowfall/pull/106/files#r580839679